src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -0,0 +1,142 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Runtime.InteropServices;
+
+namespace System.Reflection.Emit
+{
+    internal sealed class ModuleBuilderImpl : ModuleBuilder
+    {
+        private readonly Assembly _coreAssembly;
+        private readonly string _name;
+        private readonly Dictionary<string, Type> _coreTypes = new();
+
+        #region Internal Data Members
+
+        internal readonly Dictionary<Assembly, AssemblyReferenceHandle> _assemblyRefStore = new();
+        internal readonly Dictionary<Type, TypeReferenceHandle> _typeRefStore = new();
+        internal readonly List<TypeBuilderImpl> _typeDefStore = new();
+        internal int _nextMethodDefRowId = 1;
+        internal int _nextFieldDefRowId = 1;
+
+        #endregion
+
+        internal ModuleBuilderImpl(string name, Assembly coreAssembly)
+        {
+            _coreAssembly = coreAssembly;
+            _name = name;
+        }
+
+        internal Type GetTypeFromCoreAssembly(string name)
```

Comment:
For better perf, it may be better to use Enum as the key for cache of known types.

(It can wait for follow up PR. You can just add TODO for now.)

Reply:
`System.TypeCode` does not have all necessary values. For example, it is missing IntPtr/UIntPtr.

Reply:
Added TODO